src/rez/package_test.py

```diff
@@ -718,3 +841,163 @@ def print_summary(self):
+            the Rez test has no defined ``"requires"``, this parameter is empty.
+        extra_packages (set[PackageRequest], optional):
+            Include these Rez package versions in the check, if provided.  This
+            is usually empty unless the user explicitly asks for them, e.g.
+            using ``rez-test foo --interactive bar --extra-packages thing-1``.
+
+    Raises:
+        PackageTestError:
+            If ``test_requires`` or ``extra_packages`` result in a Rez request conflict.
+
+    Returns:
+        Package or Variant:
+            Get a specific variant to return. If ``package`` has no variants
+            defined then this function returns ``package``, instead.
+
+    """
+    variants = package.variants
+
+    if not variants:
+        return package
+
+    candidates = []
+
+    for variant_index, package_requests in enumerate(variants):
+        if not _is_conflict(package_requests, test_requires, extra_packages):
+            candidates.append(package.get_variant(variant_index))
+
+    if not candidates:
+        if not extra_packages:
+            raise PackageTestError(
+                "No valid variant could match your given Rez test requires."
+            )
+
+        raise PackageTestError(
+            "No valid variant could match your given Rez test requires + extra packages."
+        )
+
+    return candidates[0]
+
+
+def _pre_test_env_commands(test_name, variant, executor):
+    """Run package.py:pre_test_commands() if present.
+
+    Args:
+        test_name (str):
+            The Rez test name which will be ran.
+        variant (Package or Variant):
+            A Rez object which may define ``pre_test_commands``.
+        executor (RexExecutor):
+            An interpreter to load commands during the ``rez-env``.
+
+    """
+    _pre_test_commands(test_name, variant, executor)
+
+    if hasattr(variant, "index"):
+        executor.setenv("REZ_TEST_VARIANT_INDEX", variant.index)
+
+
+def _pre_test_commands(test_name, variant, executor):
+    """Run package.py:pre_test_commands(), if present.
+
+    Args:
+        test_name (str):
+            The Rez test name to add to ``$REZ_CURRENT_TEST_NAME``, which is
+            useful for debugging.
+        variant (Package or Variant):
+            A Rez object which may define ``pre_test_commands``.
+        executor (RexExecutor):
+            An interpreter to load commands during the ``rez-env``.
+
+    """
+    pre_test_commands = getattr(variant, "pre_test_commands")
+
+    if not pre_test_commands:
+        with executor.reset_globals():
+            executor.setenv("REZ_CURRENT_TEST_NAME", test_name)
+
+        return
+
+    test_ns = {
+        "name": test_name
+    }
+
+    with executor.reset_globals():
+        executor.bind("this", variant)
+        executor.bind("test", RO_AttrDictWrapper(test_ns))
+        executor.execute_code(pre_test_commands)
+        executor.setenv("REZ_CURRENT_TEST_NAME", test_name)
```

Comment:
```suggestion
    with executor.reset_globals():
        executor.setenv("REZ_CURRENT_TEST_NAME", test_name)

        if not pre_test_commands:
            return

        executor.bind("this", variant)
        executor.bind("test", RO_AttrDictWrapper({"name": test_name}))
        executor.execute_code(pre_test_commands)
        executor.setenv("REZ_CURRENT_TEST_NAME", test_name)
```
This would better convey the intention.

Reply:
Thank you, I was actually really hoping someone more familiar with executor would suggest a refactor. I was trying to match the original but it ended up becoming a bit weird, as you can see.